File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_REPORTVERSION.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_REPORTVERSION : Message
    {
        public enum ResultIds : UInt32
        {
            Failed = 0,
            OldGame = 1,
            Success = 2,
            Reinstall = 3,
        }

        public SID_REPORTVERSION()
        {
            Id = (byte)MessageIds.SID_REPORTVERSION;
            Buffer = new byte[6];
        }

        public SID_REPORTVERSION(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_REPORTVERSION;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] SID_REPORTVERSION ({4 + Buffer.Length} bytes)");

            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        if (Buffer.Length < 21)
                            throw new GameProtocolViolationException(context.Client, "SID_REPORTVERSION buffer must be at least 21 bytes");

                        using var m = new MemoryStream(Buffer);
                        using var r = new BinaryReader(m);

                        context.Client.GameState.Platform = (Platform.PlatformCode)r.ReadUInt32();
                        context.Client.GameState.Product = (Product.ProductCode)r.ReadUInt32();
                        context.Client.GameState.Version.VersionByte = r.ReadUInt32();
                        context.Client.GameState.Version.EXERevision = r.ReadUInt32();
                        context.Client.GameState.Version.EXEChecksum = r.ReadUInt32();
                        context.Client.GameState.Version.EXEInformation = r.ReadString();

                        return new SID_REPORTVERSION().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient));
                    }
                case MessageDirection.ServerToClient:
                    {
                        /**
                         * (UINT32) Result
                         * (STRING) Filename
                         *  (UINT8) Unknown (0)
                         */

                        var filename = "ver-IX86-1.mpq";

                        Buffer = new byte[6 + Encoding.UTF8.GetByteCount(filename)];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write((UInt32)ResultIds.Success);
                        w.Write((string)filename);
                        w.Write((byte)0); // unknown

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] SID_REPORTVERSION ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_REPORTVERSION.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_REPORTVERSION
        Inherits Message

        Public Enum ResultIds As UInt32
            Failed = 0
            OldGame = 1
            Success = 2
            Reinstall = 3
        End Enum

        Public Sub New()
            Id = CByte(MessageIds.SID_REPORTVERSION)
            Buffer = New Byte(5) {}
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_REPORTVERSION)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")

            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    If Buffer.Length < 21 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 21 bytes")

                    Using m = New MemoryStream(Buffer)
                        Using r = New BinaryReader(m)
                            context.Client.GameState.Platform = CType(r.ReadUInt32(), Platform.PlatformCode)
                            context.Client.GameState.Product = CType(r.ReadUInt32(), Product.ProductCode)
                            context.Client.GameState.Version.VersionByte = r.ReadUInt32()
                            context.Client.GameState.Version.EXERevision = r.ReadUInt32()
                            context.Client.GameState.Version.EXEChecksum = r.ReadUInt32()
                            context.Client.GameState.Version.EXEInformation = r.ReadByteString()
                        End Using
                    End Using

                    Return New SID_REPORTVERSION().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient))
                Case MessageDirection.ServerToClient
                    Dim filename = "ver-IX86-1.mpq"
                    Buffer = New Byte(6 + Encoding.UTF8.GetByteCount(filename) - 1) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(CUInt(ResultIds.Success))
                            w.Write(CStr(filename))
                            w.Write(CByte(0))
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
